framework/auth/campaigns.py

```diff
@@ -1,68 +1,134 @@
-import httplib as http
-
+from datetime import datetime, timedelta
 import furl
-from werkzeug.datastructures import ImmutableDict
-from framework.exceptions import HTTPError
+import logging
+
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound, QueryException, ImproperConfigurationError
 
 from website import mails
-from website.settings import DOMAIN
-
-CAMPAIGNS = ImmutableDict({
-    'prereg': {
-        'system_tag': 'prereg_challenge_campaign',
-        'redirect_url': lambda: furl.furl(DOMAIN).add(path='prereg/').url,
-        'confirmation_email_template': mails.CONFIRM_EMAIL_PREREG,
-    },
-    'institution': {
-        'system_tag': 'institution_campaign',
-        'redirect_url': lambda: ''
-    },
-    'erpc': {
-        'system_tag': 'erp_challenge_campaign',
-        'redirect_url': lambda: furl.furl(DOMAIN).add(path='erpc/').url,
-        'confirmation_email_template': mails.CONFIRM_EMAIL_ERPC,
-    },
-    # Various preprint services
-    # Each preprint service will offer their own campaign with appropriate distinct branding
-    'osf-preprints': {
-        'system_tag': 'osf_preprints',
-        'redirect_url': lambda: furl.furl(DOMAIN).add(path='preprints/').url,
-        'confirmation_email_template': mails.CONFIRM_EMAIL_PREPRINTS_OSF,
-        'proxy_login': True,
-    }
-})
+from website.models import PreprintProvider
+from website.settings import DOMAIN, CAMPAIGN_REFRESH_THRESHOLD
+
+
+CAMPAIGNS = None
+CAMPAIGNS_LAST_REFRESHED = datetime.utcnow()
+
+
+def get_campaigns():
```

Comment:
Instead of assessing `campaigns.CAMPAIGN`, all helper method called by OSF (such as `campaing_url_for`, `campaing_for_user`, etc.) calls `get_campaigns()` first. This will make sure that campaigns are fully initiated and updated regularly. The current threshold is 5 minutes.